File: app/src/main/java/com/xeniac/warrantyroster/mainactivity/changeemailfragment/ChangeEmailFragment.java
package com.xeniac.warrantyroster.mainactivity.changeemailfragment;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.request.RequestHeaders;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyroster.Constants;
import com.xeniac.warrantyroster.NetworkHelper;
import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.databinding.FragmentChangeEmailBinding;
import com.xeniac.warrantyroster.mainactivity.MainActivity;

import java.util.Arrays;

public class ChangeEmailFragment extends Fragment {

    private FragmentChangeEmailBinding changeEmailBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;
    private String currentEmail;

    public ChangeEmailFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        changeEmailBinding = FragmentChangeEmailBinding.inflate(inflater, container, false);
        view = changeEmailBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        changeEmailBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);
        ((MainActivity) context).hideNavBar();

        textInputsBackgroundColor();
        textInputsStrokeColor();
        returnToMainActivity();
        getAccountEmail();
        changeEmailOnClick();
        changeEmailActionDone();
    }

    private void textInputsBackgroundColor() {
        changeEmailBinding.tiChangeEmailEditEmail.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                changeEmailBinding.tiChangeEmailLayoutEmail.setBoxBackgroundColorResource(R.color.background);
            } else {
                changeEmailBinding.tiChangeEmailLayoutEmail.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });
    }

    private void textInputsStrokeColor() {
        changeEmailBinding.tiChangeEmailEditEmail.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence inputEmail, int start, int before, int count) {
                changeEmailBinding.tiChangeEmailLayoutEmail.setErrorEnabled(false);
                changeEmailBinding.tiChangeEmailLayoutEmail.setBoxStrokeColor(context.getResources().getColor(R.color.blue));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void returnToMainActivity() {
        changeEmailBinding.toolbarChangeEmail.setNavigationOnClickListener(view ->
                activity.onBackPressed());
    }

    private void getAccountEmail() {
        if (getArguments() != null) {
            ChangeEmailFragmentArgs args = ChangeEmailFragmentArgs.fromBundle(getArguments());
            currentEmail = args.getAccountEmail();
        }
    }

    private void changeEmailOnClick() {
        changeEmailBinding.btnChangeEmail.setOnClickListener(view ->
                changeUserEmail());
    }

    private void changeEmailActionDone() {
        changeEmailBinding.tiChangeEmailEditEmail.setOnEditorActionListener((textView, actionId, keyEvent) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                changeUserEmail();
            }
            return false;
        });
    }

    private void changeUserEmail() {
        InputMethodManager inputMethodManager = (InputMethodManager)
                context.getSystemService(Context.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(view.getApplicationWindowToken(), 0);

        if (NetworkHelper.hasNetworkAccess(context)) {
            getChangeUserEmailInputs();
        } else {
            hideLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> changeUserEmail()).show();
        }
    }

    private void getChangeUserEmailInputs() {
        String newEmail;

        if (TextUtils.isEmpty(changeEmailBinding.tiChangeEmailEditEmail.getText())) {
            changeEmailBinding.tiChangeEmailLayoutEmail.requestFocus();
            changeEmailBinding.tiChangeEmailLayoutEmail.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else {
            newEmail = changeEmailBinding.tiChangeEmailEditEmail.getText().toString().trim().toLowerCase();

            if (!isEmailValid(newEmail)) {
                changeEmailBinding.tiChangeEmailLayoutEmail.requestFocus();
                changeEmailBinding.tiChangeEmailLayoutEmail.setError(context.getResources().getString(R.string.change_email_error_email));
            } else if (newEmail.equals(currentEmail)) {
                changeEmailBinding.tiChangeEmailLayoutEmail.requestFocus();
                changeEmailBinding.tiChangeEmailLayoutEmail.setError(context.getResources().getString(R.string.change_email_error_email_same));
            } else {
                changeUserEmailMutation(newEmail);
            }
        }
    }

    private void changeUserEmailMutation(String newEmail) {
        showLoadingAnimation();

        SharedPreferences loginPrefs = context.getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE);
        String userToken = loginPrefs.getString(Constants.PREFERENCE_USER_TOKEN_KEY, null);

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new ChangeUserEmailMutation(newEmail))
                .toBuilder().requestHeaders(RequestHeaders.builder().addHeader("Authorization", "Bearer " + userToken).build())
                .build()
                .enqueue(new ApolloCall.Callback<ChangeUserEmailMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<ChangeUserEmailMutation.Data> response) {
                        if (changeEmailBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("changeUserEmail", "onResponse: " + response);
                                    MaterialAlertDialogBuilder changeUserEmailDialogBuilder = new MaterialAlertDialogBuilder(context)
                                            .setMessage(context.getResources().getString(R.string.change_email_dialog_message))
                                            .setPositiveButton(context.getResources().getString(R.string.change_email_dialog_positive), (dialog, which) -> {
                                            })
                                            .setOnDismissListener(dialog -> activity.onBackPressed());
                                    changeUserEmailDialogBuilder.show();
                                } else {
                                    Log.e("changeUserEmail", "onResponse Errors: " + response.getErrors());
                                    if (Arrays.toString(response.getErrors().get(0).getCustomAttributes().values().toArray()).contains("already exists")) {
                                        Snackbar.make(view, context.getResources().getString(R.string.change_email_error_email_exists),
                                                BaseTransientBottomBar.LENGTH_LONG)
                                                .show();
                                    } else {
                                        Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                                BaseTransientBottomBar.LENGTH_INDEFINITE)
                                                .setAction(context.getResources().getString(R.string.network_error_retry), v -> changeUserEmail()).show();
                                    }
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("changeUserEmail", "onFailure: " + e.getMessage());
                        if (changeEmailBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void showLoadingAnimation() {
        changeEmailBinding.tiChangeEmailEditEmail.setEnabled(false);
        changeEmailBinding.btnChangeEmail.setClickable(false);
        changeEmailBinding.btnChangeEmail.setText(null);
        changeEmailBinding.cpiChangeEmail.setVisibility(View.VISIBLE);
    }

    private void hideLoadingAnimation() {
        changeEmailBinding.cpiChangeEmail.setVisibility(View.GONE);
        changeEmailBinding.tiChangeEmailEditEmail.setEnabled(true);
        changeEmailBinding.btnChangeEmail.setClickable(true);
        changeEmailBinding.btnChangeEmail.setText(context.getResources().getString(R.string.change_email_btn_change));
    }

    private boolean isEmailValid(String email) {
        return Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }
}

File: app/src/main/java/com/xeniac/warrantyroster_manager/mainactivity/changeemailfragment/ChangeEmailFragment.kt
package com.xeniac.warrantyroster_manager.mainactivity.changeemailfragment

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.util.Patterns
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.xeniac.warrantyroster_manager.util.NetworkHelper
import com.xeniac.warrantyroster_manager.R
import com.xeniac.warrantyroster_manager.databinding.FragmentChangeEmailBinding
import com.xeniac.warrantyroster_manager.mainactivity.MainActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

class ChangeEmailFragment : Fragment(R.layout.fragment_change_email) {

    private var _binding: FragmentChangeEmailBinding? = null
    private val binding get() = _binding!!

    private val firebaseAuth = FirebaseAuth.getInstance()
    private val currentUser = firebaseAuth.currentUser

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentChangeEmailBinding.bind(view)
        (requireContext() as MainActivity).hideNavBar()

        textInputsBackgroundColor()
        textInputsStrokeColor()
        returnToMainActivity()
        changeEmailOnClick()
        changeEmailActionDone()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun textInputsBackgroundColor() {
        binding.tiEditPassword.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiLayoutPassword.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiLayoutPassword.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiEditNewEmail.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiLayoutNewEmail.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiLayoutNewEmail.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }
    }

    private fun textInputsStrokeColor() {
        binding.tiEditPassword.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputPassword: CharSequence?, start: Int, before: Int, count: Int
            ) {
                binding.tiLayoutPassword.boxStrokeColor =
                    ContextCompat.getColor(requireContext(), R.color.blue)
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })

        binding.tiEditNewEmail.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputEmail: CharSequence?, start: Int, before: Int, count: Int
            ) {
                binding.tiLayoutNewEmail.isErrorEnabled = false
                binding.tiLayoutNewEmail.boxStrokeColor =
                    ContextCompat.getColor(requireContext(), R.color.blue)
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })
    }

    private fun returnToMainActivity() = binding.toolbar.setNavigationOnClickListener {
        requireActivity().onBackPressed()
    }

    private fun changeEmailOnClick() = binding.btnChangeEmail.setOnClickListener {
        changeUserEmail()
    }

    private fun changeEmailActionDone() =
        binding.tiEditNewEmail.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                changeUserEmail()
            }
            false
        }

    private fun changeUserEmail() {
        val inputMethodManager = requireContext()
            .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputMethodManager.hideSoftInputFromWindow(binding.root.applicationWindowToken, 0)

        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            getChangeUserEmailInputs()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { changeUserEmail() }
                show()
            }
        }
    }

    private fun getChangeUserEmailInputs() {
        val password = binding.tiEditPassword.text.toString().trim()
        val newEmail = binding.tiEditNewEmail.text.toString().trim().lowercase()

        if (password.isBlank()) {
            binding.tiLayoutPassword.requestFocus()
            binding.tiLayoutPassword.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else if (newEmail.isBlank()) {
            binding.tiLayoutNewEmail.requestFocus()
            binding.tiLayoutNewEmail.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else {
            if (!isEmailValid(newEmail)) {
                binding.tiLayoutNewEmail.requestFocus()
                binding.tiLayoutNewEmail.error =
                    requireContext().getString(R.string.change_email_error_new_email)
            } else if (newEmail == currentUser?.email) {
                binding.tiLayoutNewEmail.requestFocus()
                binding.tiLayoutNewEmail.error =
                    requireContext().getString(R.string.change_email_error_email_same)
            } else {
                showLoadingAnimation()
                reAuthenticateUser(password, newEmail)
            }
        }
    }

    private fun reAuthenticateUser(password: String, newEmail: String) =
        CoroutineScope(Dispatchers.IO).launch {
            try {
                currentUser?.let {
                    val credential = EmailAuthProvider.getCredential(it.email.toString(), password)
                    it.reauthenticate(credential).await()
                    Log.i("reAuthenticateUser", "User re-authenticated.")
                    changeUserEmailAuth(newEmail)
                }
            } catch (e: Exception) {
                Log.e("reAuthenticateUser", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    when {
                        e.toString()
                            .contains("The password is invalid or the user does not have a password") -> {
                            Snackbar.make(
                                binding.root,
                                requireContext().getString(R.string.change_email_error_credentials),
                                LENGTH_INDEFINITE
                            ).show()
                        }
                        else -> {
                            Snackbar.make(
                                binding.root,
                                requireContext().getString(R.string.network_error_failure),
                                LENGTH_LONG
                            ).show()
                        }
                    }
                }
            }
        }

    private fun changeUserEmailAuth(newEmail: String) =
        CoroutineScope(Dispatchers.IO).launch {
            try {
                currentUser?.let {
                    it.updateEmail(newEmail).await()
                    Log.i("changeUserEmailAuth", "User email updated to ${newEmail}.")

                    withContext(Dispatchers.Main) {
                        hideLoadingAnimation()
                        MaterialAlertDialogBuilder(requireContext()).apply {
                            setMessage(requireContext().getString(R.string.change_email_dialog_message))
                            setPositiveButton(requireContext().getString(R.string.change_email_dialog_positive)) { _, _ -> }
                            setOnDismissListener { requireActivity().onBackPressed() }
                            show()
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("changeUserEmailAuth", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    when {
                        e.toString()
                            .contains("The email address is already in use by another account") -> {
                            Snackbar.make(
                                binding.root,
                                requireContext().getString(R.string.change_email_error_email_exists),
                                LENGTH_LONG
                            ).show()
                        }
                        else -> {
                            Snackbar.make(
                                binding.root,
                                requireContext().getString(R.string.network_error_failure),
                                LENGTH_LONG
                            ).show()
                        }
                    }
                }
            }
        }

    private fun showLoadingAnimation() {
        binding.tiEditPassword.isEnabled = false
        binding.tiEditNewEmail.isEnabled = false
        binding.btnChangeEmail.isClickable = false
        binding.btnChangeEmail.text = null
        binding.cpiChangeEmail.visibility = VISIBLE
    }

    private fun hideLoadingAnimation() {
        binding.cpiChangeEmail.visibility = GONE
        binding.tiEditPassword.isEnabled = true
        binding.tiEditNewEmail.isEnabled = true
        binding.btnChangeEmail.isClickable = true
        binding.btnChangeEmail.text = requireContext().getString(R.string.change_email_btn_change)
    }

    private fun isEmailValid(email: String): Boolean =
        Patterns.EMAIL_ADDRESS.matcher(email).matches()
}